File: server/internal/protocol/protocol.go
// Package protocol defines the JSON message contracts exchanged between
// clients and the Mage Craft authoritative server over WebSocket (see the
// project plan, §4). Every message is a flat JSON object carrying its own
// "type" discriminator, so a caller can inspect the type with PeekType
// before decoding into the matching Go struct.
package protocol

import (
	"encoding/json"
	"fmt"
)

// Client -> server message types.
const (
	TypeCreateRoom    = "create_room"
	TypeJoinRoom      = "join_room"
	TypeListRooms     = "list_rooms"
	TypeSelectTeam    = "select_team"
	TypeSelectElement = "select_element"
	TypeAddBot        = "add_bot"
	TypeRemoveBot     = "remove_bot"
	TypeClaimSlot     = "claim_slot"
	TypeSetReady      = "set_ready"
	TypeStartMatch    = "start_match"
	TypeInput         = "input"
)

// Server -> client message types.
const (
	TypeRoomState  = "room_state"
	TypeRoomList   = "room_list"
	TypeMatchStart = "match_start"
	TypeSnapshot   = "snapshot"
	TypeRoundEnd   = "round_end"
	TypeError      = "error"
)

// Vec2DTO is the wire representation of a 2D vector (position, direction or
// velocity). It intentionally avoids depending on the game package's own
// vector type so protocol stays decoupled from the simulation internals.
type Vec2DTO struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// typeEnvelope is used only to peek at the "type" discriminator of an
// otherwise-unknown incoming message.
type typeEnvelope struct {
	Type string `json:"type"`
}

// PeekType reports the "type" field of a raw client/server message without
// decoding the rest of it, so the caller can dispatch to the right concrete
// struct. It returns an error if data isn't valid JSON or has no (or an
// empty) "type" field.
func PeekType(data []byte) (string, error) {
	var env typeEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return "", fmt.Errorf("protocol: invalid message: %w", err)
	}
	if env.Type == "" {
		return "", fmt.Errorf("protocol: message is missing a \"type\" field")
	}
	return env.Type, nil
}

// --- Client -> server messages ---

// CreateRoomMsg requests a new room with the given team size (1-6, i.e. up
// to 6x6, see GDD §7). When FillBots is true the server will fill remaining
// seats with bots after humans take their slots (typically via FillEmptyWithBots).
type CreateRoomMsg struct {
	Type          string `json:"type"`
	TeamSize      int    `json:"teamSize"`
	FillBots      bool   `json:"fillBots,omitempty"`
	BotDifficulty string `json:"botDifficulty,omitempty"` // easy|normal|hard; default normal
}

// JoinRoomMsg requests joining an existing room by its short room code.
// In lobby the player joins as a seat candidate; during in_progress they
// join as a spectator until rematch.
type JoinRoomMsg struct {
	Type   string `json:"type"`
	RoomID string `json:"roomId"`
	Name   string `json:"name"`
}

// ListRoomsMsg asks the server for the current public room catalog.
type ListRoomsMsg struct {
	Type string `json:"type"`
}

// SelectTeamMsg picks team 0 or 1 within the current room.
type SelectTeamMsg struct {
	Type string `json:"type"`
	Team int    `json:"team"`
}

// SelectElementMsg picks one of the 7 catalog elements (GDD §8.1) for the
// sender's mage. The server rejects it if another mage on the same team
// already has that element (GDD §7 uniqueness rule).
type SelectElementMsg struct {
	Type    string `json:"type"`
	Element string `json:"element"`
}

// AddBotMsg asks the host to fill an empty slot on a team with a bot of the
// given difficulty ("easy" | "normal" | "hard"). The bot auto-picks a free
// element on that team.
type AddBotMsg struct {
	Type       string `json:"type"`
	Team       int    `json:"team"`
	Difficulty string `json:"difficulty"`
}

// RemoveBotMsg removes a previously added bot, freeing its slot and element.
type RemoveBotMsg struct {
	Type   string `json:"type"`
	SlotID string `json:"slotId"`
}

// ClaimSlotMsg lets a spectator reserve a bot (or empty) slot for the next
// rematch round. The bot keeps playing until round_end.
type ClaimSlotMsg struct {
	Type   string `json:"type"`
	SlotID string `json:"slotId"`
}

// SetReadyMsg toggles the sender's ready state in the lobby.
type SetReadyMsg struct {
	Type  string `json:"type"`
	Ready bool   `json:"ready"`
}

// StartMatchMsg asks the room to start the match. The server validates that
// every slot is filled and every mage has a valid element before accepting.
type StartMatchMsg struct {
	Type string `json:"type"`
}

// InputMsg carries one input sample for the sender's mage during a match.
type InputMsg struct {
	Type     string  `json:"type"`
	Move     Vec2DTO `json:"move"`
	Aim      Vec2DTO `json:"aim"`
	Charging bool    `json:"charging"`
	Release  bool    `json:"release"`
}

// --- Server -> client messages ---

// PlayerSlotDTO describes one slot (human or bot) inside a room, as shown to
// clients in RoomStateMsg.
type PlayerSlotDTO struct {
	SlotID               string `json:"slotId"`
	Team                 int    `json:"team"`
	PlayerID             string `json:"playerId,omitempty"`
	Name                 string `json:"name,omitempty"`
	IsBot                bool   `json:"isBot"`
	Element              string `json:"element,omitempty"`
	Ready                bool   `json:"ready"`
	PendingClaimPlayerID string `json:"pendingClaimPlayerId,omitempty"`
}

// SpectatorDTO is a human watching an in-progress match (or waiting in
// rematch lobby after claiming a slot that has not yet been applied).
type SpectatorDTO struct {
	PlayerID      string `json:"playerId"`
	Name          string `json:"name"`
	ClaimedSlotID string `json:"claimedSlotId,omitempty"`
}

// RoomStateMsg is broadcast to everyone in a room whenever the lobby state
// changes (join/leave/select_team/select_element/add_bot/remove_bot/ready/claim).
type RoomStateMsg struct {
	Type       string          `json:"type"`
	RoomID     string          `json:"roomId"`
	TeamSize   int             `json:"teamSize"`
	State      string          `json:"state"` // lobby | in_progress | ended
	FillBots   bool            `json:"fillBots,omitempty"`
	Slots      []PlayerSlotDTO `json:"slots"`
	Spectators []SpectatorDTO  `json:"spectators,omitempty"`
	YouRole    string          `json:"youRole,omitempty"` // player | spectator (per-recipient)
}

// RoomSummaryDTO is one entry in a room_list response.
type RoomSummaryDTO struct {
	RoomID            string `json:"roomId"`
	TeamSize          int    `json:"teamSize"`
	State             string `json:"state"`
	Filled            int    `json:"filled"`
	Capacity          int    `json:"capacity"`
	OpenBotSlots      int    `json:"openBotSlots"`
	AcceptsSpectators bool   `json:"acceptsSpectators"`
}

// RoomListMsg is the catalog of joinable rooms (lobby + in_progress).
type RoomListMsg struct {
	Type  string           `json:"type"`
	Rooms []RoomSummaryDTO `json:"rooms"`
}

// MatchStartMsg tells clients the match's 60Hz simulation loop has started.
type MatchStartMsg struct {
	Type string `json:"type"`
}

// MageSnapshotDTO is one mage's state in a simulation snapshot.
type MageSnapshotDTO struct {
	ID       string  `json:"id"`
	Team     int     `json:"team"`
	Position Vec2DTO `json:"position"`
	Facing   Vec2DTO `json:"facing"`
	Health   float64 `json:"health"`
	Lives    int     `json:"lives"`
	Charging bool    `json:"charging"`
	Charge   float64 `json:"charge"`
	Element  string  `json:"element"`
}

// ProjectileSnapshotDTO is one in-flight projectile's state in a snapshot.
type ProjectileSnapshotDTO struct {
	ID       string  `json:"id"`
	Element  string  `json:"element"`
	Position Vec2DTO `json:"position"`
	Velocity Vec2DTO `json:"velocity"`
}

// PuddleSnapshotDTO is one active ground puddle's state in a snapshot
// (GDD §8.5).
type PuddleSnapshotDTO struct {
	ID        string  `json:"id"`
	Position  Vec2DTO `json:"position"`
	Radius    float64 `json:"radius"`
	Remaining float64 `json:"remaining"`
}

// SnapshotMsg is broadcast at the server's snapshot rate (~20-30Hz) while a
// match is in progress.
type SnapshotMsg struct {
	Type        string                  `json:"type"`
	Tick        uint64                  `json:"tick"`
	Mages       []MageSnapshotDTO       `json:"mages"`
	Projectiles []ProjectileSnapshotDTO `json:"projectiles"`
	Puddles     []PuddleSnapshotDTO     `json:"puddles"`
}

// RoundEndMsg announces the winning team once a round ends (opposing team
// reaches zero lives). After this the server resets the room to lobby for rematch.
type RoundEndMsg struct {
	Type       string `json:"type"`
	WinnerTeam int    `json:"winnerTeam"`
}

// ErrorMsg reports a rejected action (e.g. invalid element, room full) back
// to the client that caused it.
type ErrorMsg struct {
	Type    string `json:"type"`
	Message string `json:"message"`
}

File: sim/protocol.ts
/**
 * The client<->server wire contract (JSON over WebSocket).
 *
 * This file is *the* definition, imported by both sides: the client through
 * `src/net/protocol.ts` and the server through `server/src/App.ts`. It used to
 * be a TypeScript file that mirrored `server/internal/protocol/protocol.go` by
 * hand — the two are now one file, so a message can no longer change shape on
 * one side only.
 */

export type Vec2DTO = { x: number; y: number };

/**
 * Strips a `Vec2` down to the wire shape. Lives here rather than beside the
 * sender because every producer of a snapshot needs it — the server, and any
 * client running a match locally.
 */
export function fromVec2(v: { x: number; y: number }): Vec2DTO {
  return { x: v.x, y: v.y };
}

export type CreateRoomMsg = {
  type: 'create_room';
  teamSize: number;
  fillBots?: boolean;
  botDifficulty?: string;
};

export type JoinRoomMsg = { type: 'join_room'; roomId: string; name: string };
export type ListRoomsMsg = { type: 'list_rooms' };
export type SelectTeamMsg = { type: 'select_team'; team: number };
export type SelectElementMsg = { type: 'select_element'; element: string };
export type AddBotMsg = { type: 'add_bot'; team: number; difficulty: string };
export type RemoveBotMsg = { type: 'remove_bot'; slotId: string };
export type ClaimSlotMsg = { type: 'claim_slot'; slotId: string };
export type SetReadyMsg = { type: 'set_ready'; ready: boolean };
export type StartMatchMsg = { type: 'start_match' };

/**
 * Matchmaking (Clash Royale-style): the player presses Battle, waits in a
 * queue, and the server builds the 1v1 room. This is the primary way into a
 * match; `create_room`/`join_room` survive for private/custom games.
 */
export type JoinQueueMsg = {
  type: 'join_queue';
  name: string;
  /** @deprecated Send `set_loadout` instead — it also covers custom rooms. */
  deck?: string[];
  /** Current Elo, so a paired opponent can be told what they're facing. */
  rating?: number;
};
export type LeaveQueueMsg = { type: 'leave_queue' };

/**
 * What the player brings to their next match: the 8-card deck, the 4-mage
 * squad, and — since the idle pivot — the strategy program that plays the deck
 * for them (GDD §7). Sent once after connecting, before joining a queue or a
 * room, because it has to apply to both paths — a room seat is claimed long
 * after any queue message would have been sent.
 *
 * Any part may be omitted, in which case the server keeps its default.
 *
 * `strategy` is deliberately `unknown` rather than `Strategy`: it is a nested
 * program authored on an untrusted client, so the wire type promises nothing
 * and `validateStrategy` on the server is what decides whether it is one.
 */
export type SetLoadoutMsg = {
  type: 'set_loadout';
  deck?: string[];
  squad?: string[];
  strategy?: unknown;
};

/**
 * Spend mana to cast a spell at a chosen point.
 *
 * @deprecated Since the idle pivot nobody casts by hand — a seat is played by
 * the `Tactician` running the player's program, and the server rejects this
 * message. It is kept because `MatchTransport` implements it on both sides and
 * it is the natural hook for a future override mode; delete it only once
 * something has replaced that seam.
 */
export type CastMsg = { type: 'cast'; cardId: string; position: Vec2DTO };

/** A quick-react from the hand, GDD-adjacent but not in it — see EmoteId in src/app/emotes.ts. */
export type SendEmoteMsg = { type: 'send_emote'; emoteId: string };

export type ClientMsg =
  | CreateRoomMsg
  | JoinRoomMsg
  | ListRoomsMsg
  | SelectTeamMsg
  | SelectElementMsg
  | AddBotMsg
  | RemoveBotMsg
  | ClaimSlotMsg
  | SetReadyMsg
  | StartMatchMsg
  | JoinQueueMsg
  | LeaveQueueMsg
  | SetLoadoutMsg
  | CastMsg
  | SendEmoteMsg;

export type PlayerSlotDTO = {
  slotId: string;
  team: number;
  playerId?: string;
  name?: string;
  isBot: boolean;
  element?: string;
  ready: boolean;
  pendingClaimPlayerId?: string;
};

export type SpectatorDTO = {
  playerId: string;
  name: string;
  claimedSlotId?: string;
};

export type RoomStateMsg = {
  type: 'room_state';
  roomId: string;
  teamSize: number;
  state: string;
  fillBots?: boolean;
  slots: PlayerSlotDTO[];
  spectators?: SpectatorDTO[];
  youRole?: string;
};

export type RoomSummaryDTO = {
  roomId: string;
  teamSize: number;
  state: string;
  filled: number;
  capacity: number;
  openBotSlots: number;
  acceptsSpectators: boolean;
};

export type RoomListMsg = { type: 'room_list'; rooms: RoomSummaryDTO[] };

/** One status effect on the wire: `k` is the {@link EffectKind}, `s` the stacks. */
export type MageFxDTO = { k: string; s?: number };

export type MageSnapshotDTO = {
  id: string;
  team: number;
  position: Vec2DTO;
  facing: Vec2DTO;
  health: number;
  /** Per-unit since the pivot — a Golem and an Archer do not share a max. */
  maxHealth: number;
  charging: boolean;
  charge: number;
  element: string;
  /** Identity: 'tank' | 'damage' | 'support' (GDD §8). */
  role: string;
  /** Which roster entry this mage is, omitted for legacy/bare mages. */
  rosterId?: string;
  /**
   * Status effects running on this mage (GDD §9), in `EFFECT_ORDER`. Omitted
   * entirely when there are none, which is the common case. `s` is the stack
   * count, omitted when 1.
   *
   * This replaced the `shielded`/`hasted`/`slowed` booleans. Those cost three
   * fields per mage per snapshot whether or not anything was happening, and
   * every new effect cost a protocol change; a list costs nothing when idle and
   * never needs another field.
   */
  fx?: MageFxDTO[];
  /**
   * Enemy mages this one has personally put down (GDD §4). A team's total is
   * *not* the sum of these — see `World.kill` — it is the opposing team's
   * `deaths`, which is what the squad panel adds up.
   */
  kills: number;
  /** Times this mage has been put down; survives respawn. */
  deaths: number;
  /** Seconds until this mage returns. Omitted while it is alive. */
  respawnRemaining?: number;
  /** Post-respawn damage immunity is running. Omitted when false. */
  immune?: boolean;
};

export type StructureSnapshotDTO = {
  id: string;
  team: number;
  /** 'core' | 'tower' */
  kind: string;
  position: Vec2DTO;
  radius: number;
  health: number;
  maxHealth: number;
  alive: boolean;
  /** A Core reads as immune while its Towers stand (GDD §5). */
  invulnerable: boolean;
};

export type ProjectileSnapshotDTO = {
  id: string;
  element: string;
  position: Vec2DTO;
  velocity: Vec2DTO;
  /**
   * Height above the ground plane. Without it the client drew every spell
   * sliding at y=0, half-buried in the ground and casting no shadow — the arc
   * a charged throw actually flies only exists here.
   */
  height: number;
  /** The projectile's real collision radius, so a boulder reads bigger than a bolt. */
  radius: number;
};

/**
 * A spell that was just cast, for VFX only (GDD §17). Buffs and curses land
 * instantly on the mages in range, so without this the wire would show no
 * trace of three of the four spells in the deck.
 */
export type SpellCastDTO = {
  id: string;
  /** 'blessing' | 'slow_curse' | 'arcane_shield' | 'plague'. */
  spellId: string;
  team: number;
  position: Vec2DTO;
  radius: number;
};

export type PuddleSnapshotDTO = {
  id: string;
  position: Vec2DTO;
  radius: number;
  remaining: number;
};

/**
 * The rule that last put a spell down for the receiving player (GDD §7).
 *
 * This is the whole of an idle player's in-match feedback. They did not click,
 * so without being told *which of their own rules* caused what they are
 * watching, a match is something that happens near them rather than something
 * they wrote — which is the difference between this game and a screensaver.
 *
 * Sent on the per-recipient channel next to `mana`/`hand`, never broadcast: a
 * player's program is theirs, and leaking the opponent's would turn reading
 * the wire into a strategy dump.
 */
export type FiredRuleDTO = {
  ruleId: string;
  /** 0-based position in the program; the HUD says "Regra {index + 1}". */
  index: number;
  cardId: string;
  /** The `TargetSelector` the rule named — where it aimed, in the player's words. */
  at: string;
};

export type SnapshotMsg = {
  type: 'snapshot';
  tick: number;
  mages: MageSnapshotDTO[];
  projectiles: ProjectileSnapshotDTO[];
  puddles: PuddleSnapshotDTO[];
  structures: StructureSnapshotDTO[];
  /** Casts still inside their VFX window; a client fires each one once, by id. */
  spells: SpellCastDTO[];
  /** Seconds of match time elapsed, for the countdown (GDD §4). */
  elapsed: number;
  suddenDeath: boolean;
  /** The receiving player's own mana, 0..MANA_MAX (GDD §6). */
  mana: number;
  /**
   * The receiving player's own hand, in slot order (GDD §7). The server owns the
   * deck and its cycle, so the client cannot derive this — it is sent every
   * snapshot rather than as a separate event, which also makes a rejected cast
   * self-correcting. Empty for a spectator.
   */
  hand: string[];
  /** The card entering the hand next; omitted when the deck cannot say yet. */
  next?: string;
  /**
   * The receiving player's last rule to actually cast. Omitted until one has —
   * an empty program never sets it, which is exactly what an empty program
   * should look like on screen.
   */
  firedRule?: FiredRuleDTO;
};

export type MatchStartMsg = { type: 'match_start' };
/** `winnerTeam` is -1 on a draw (GDD §4 — sudden death can expire level). */
export type RoundEndMsg = { type: 'round_end'; winnerTeam: number };
export type ErrorMsg = { type: 'error'; message: string };

export type MageStatDTO = { rosterId?: string; role: string; kills: number; deaths: number };
export type CastStatDTO = { cardId: string; casts: number };
export type TeamSummaryDTO = {
  squad: string[];
  kills: number;
  deaths: number;
  structuresDestroyed: number;
  casts: CastStatDTO[];
  mages: MageStatDTO[];
};

/**
 * The match's account of itself, sent alongside `round_end`.
 *
 * Kept as its own message rather than fields on `RoundEndMsg` because
 * `round_end` drives the rematch-lobby handover on the client, and a consumer
 * that only wants the numbers (history, analytics, a future replay) should not
 * have to touch that path.
 */
export type MatchResultMsg = {
  type: 'match_result';
  winnerTeam: number;
  durationSeconds: number;
  suddenDeath: boolean;
  /** Keyed by wire team number. */
  perTeam: Record<number, TeamSummaryDTO>;
};

/** Queue feedback: how long you have waited and how many are searching. */
export type QueueStatusMsg = {
  type: 'queue_status';
  waiting: number;
  position: number;
  elapsedSeconds: number;
};

/** Sent to both players the instant the server pairs them. */
export type MatchFoundMsg = {
  type: 'match_found';
  roomId: string;
  opponentName: string;
  yourTeam: number;
  /** True when the opponent is a bot because the queue timed out. */
  againstBot: boolean;
  /** Opponent's Elo at pairing time; null against a bot (bots aren't rated). */
  opponentRating: number | null;
};

/** Broadcast to both players in the room the instant one of them sends one — no ack, no history. */
export type EmoteMsg = { type: 'emote'; playerId: string; emoteId: string };

export type ServerMsg =
  | RoomStateMsg
  | RoomListMsg
  | MatchStartMsg
  | SnapshotMsg
  | RoundEndMsg
  | MatchResultMsg
  | QueueStatusMsg
  | MatchFoundMsg
  | EmoteMsg
  | ErrorMsg;

export function peekType(data: unknown): string | null {
  if (!data || typeof data !== 'object') return null;
  const t = (data as { type?: unknown }).type;
  return typeof t === 'string' ? t : null;
}
